### PR TITLE
Add message loop to SOA bot

### PR DESCRIPTION
**Pull request by aphedges**
This PR implements the simple messaging loop and contains some related changes, including a fix to a messaging bug I had noticed while reading through the code.

The only "regression" I noticed was that our testing time increased from ~40s to ~140s on my local machine. This is almost entirely accounted for by the messaging logic needing to be run an extra two times.

### CODE REVIEW

**Review by RunzheYang — APPROVED** [review]
LGTM!

**Comment by aphedges** [comment]
I will hold off on merging this until other changes are merged because I can rebase more easily than the others.

**Pull request merged**
